lib/tree.rb

```diff
@@ -2,59 +2,133 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
+  end
+
+  def add_helper(current, new_node)
+    return new_node if current.nil?
+
+    if new_node.key <= current.key
+      current.left = add_helper(current.left, new_node)
+    else
+      current.right = add_helper(current.right, new_node)
+    end
+    return current
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: worst case O(n)
+  # Space Complexity: worst case O(n)
   def find(key)
-    raise NotImplementedError
+    if @root.nil?
+      return nil
+    elsif @root.key == key
+      return @root.value
+    else
+      find_helper(@root, key)
+    end
+  end
+
+  def find_helper(current, key)
+    return current.value if current.key == key
+
+    if key <= current.key
+      find_helper(current.left, key)
+    else
+      find_helper(current.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  #Right, Root, Left
+  def inorder_helper(current, array)
+    return array if current.nil?
+    inorder_helper(current.left, array)
+    array << { key: current.key, value: current.value }
+    inorder_helper(current.right, array)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  #Root, Left, Right
+  def preorder_helper(current, array)
+    return array if current.nil?
+    array << { key: current.key, value: current.value }
+    preorder_helper(current.left, array)
+    preorder_helper(current.right, array)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  #Left, Right, Root
+  def postorder_helper(current, array)
+    return array if current.nil?
+
+    postorder_helper(current.left, array)
+    postorder_helper(current.right, array)
+    array << { key: current.key, value: current.value }
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def height
```

Comment:
👍 , although the space complexity is O(log n) if the tree is balanced and O(n) otherwise.